webapp/src/contexts/SessionContext.tsx

```diff
@@ -0,0 +1,226 @@
+import { ethers, type Wallet } from "ethers";
+import {
+	createContext,
+	type ReactNode,
+	useCallback,
+	useContext,
+	useEffect,
+	useMemo,
+	useState,
+	useTransition,
+} from "react";
+import { useBrowserProvider } from "@/hooks/useBrowserProvider";
+import { useHarbourRpcProvider } from "@/hooks/useRpcProvider";
+import {
+	type EncryptionKey,
+	fetchEncryptionKey,
+	type HarbourContractSettings,
+} from "@/lib/harbour";
+import {
+	deserializeSession,
+	type Session,
+	serializeSession,
+	signinToSession,
+} from "@/lib/session";
+
+const STORAGE_KEY_PREFIX = "session.v1";
+
+type Address = string;
+type Hex = string;
+
+interface SessionStorageKey {
+	chainId: bigint;
+	address: Address;
+}
+
+interface CryptoKeyPairWithHex extends CryptoKeyPair {
+	publicKeyHex: Hex;
+}
+
+interface SessionKeys {
+	encryption: CryptoKeyPairWithHex;
+	relayer: Wallet;
+}
+
+interface SessionState {
+	session: Session;
+	hasPendingRegistration: boolean;
+}
+
+interface SessionValue {
+	keys: SessionKeys | null;
+	pendingRegistration: EncryptionKey | null;
+	connect: (settings?: HarbourContractSettings) => void;
+	create: (settings?: HarbourContractSettings) => void;
+	disconnect: () => void;
+	connected: boolean;
+	isUpdating: boolean;
+	error: Error | null;
+}
+
+function storageKeyFor({ chainId, address }: SessionStorageKey): string {
+	return `${STORAGE_KEY_PREFIX}:${chainId}@${ethers.getAddress(address)}`;
+}
+
+async function loadSession(key: SessionStorageKey): Promise<Session | null> {
+	try {
+		const storageKey = storageKeyFor(key);
+		const encoded = localStorage.getItem(storageKey) ?? "";
+		return await deserializeSession(encoded);
+	} catch {
+		return null;
+	}
+}
+
+function storeSession(key: SessionStorageKey, session: Session) {
+	const storageKey = storageKeyFor(key);
+	const encoded = serializeSession(session);
+	localStorage.setItem(storageKey, encoded);
+}
+
+function onchainMatchesSession(
+	onchain: EncryptionKey,
+	{ registration }: Pick<Session, "registration">,
+) {
+	return (
+		onchain.context === registration.context &&
+		onchain.publicKey === registration.publicKey
+	);
+}
+
+function isError(o: unknown): o is Error {
+	return Object.prototype.toString.call(o) === "[object Error]";
+}
+
+const SessionContext = createContext<SessionValue | null>(null);
+
+function SessionProvider({ children }: { children: ReactNode }) {
+	const wallet = useBrowserProvider();
+	const { provider } = useHarbourRpcProvider();
+
+	const [session, setSession] = useState<SessionState | null>(null);
+	const [error, setError] = useState<Error | null>(null);
+	const [isUpdating, startUpdate] = useTransition();
+
+	const update = useCallback((handler: () => Promise<SessionState | null>) => {
+		setError(null);
+		startUpdate(async () => {
+			try {
+				const session = await handler();
+				setSession(session);
+			} catch (err) {
+				console.error(err);
+				setSession(null);
```

Comment:
Why do you set session to null here? 

I would expect that if we somehow set a session before that it is still valid, otherwise we should potentially set it to `null` when we start the update.

Reply:
It was an "extraordinary path" that isn't usually taken, and I thought it would be safer to clear the session. I will leave the session untouched (as, now that you bring it up, it does make a bit more sense that way).